### PR TITLE
Added all_metrics to CheckpointCallback, issue #431

### PR DESCRIPTION
**Pull request by Arquestro**
## Description

All metrics now will be collected in list, and written with best, n_best, last checkpoint metrics.

## Related Issue

https://github.com/catalyst-team/catalyst/issues/431

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [X] I have read the [CODE_OF_CONDUCT](../CODE_OF_CONDUCT.md) document.
- [X] I have read the [CONTRIBUTING](../CONTRIBUTING.md) document.
- [X] I have checked the code-style using `make check-style`.
- [ ] I have written the docstring in Google format for all the methods and classes that I used.
- [ ] I have checked the docs using `make check-docs`.

### CODE REVIEW

**Comment by Arquestro** [comment]
I know, i haven't made any changes to BaseCheckpointCallback here. But i wonder why do we have to do it this way, changing base class, if such logic is complex enough?

**Comment by Arquestro** [comment]
@TezRomacH What do you think about this solution?

**Comment by TezRomacH** [comment]
I'm gonna check it tonight, thanks a lot for participating ❤️

File: catalyst/dl/callbacks/checkpoint.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def __init__(
         self.resume = resume
         self.resume_dir = resume_dir
         self.top_best_metrics = []
-
+        self.all_metrics = []
```
Comment (Scitator) [inline catalyst/dl/callbacks/checkpoint.py]: why `all_metrics`? are they... `epoch_metrics`?

File: catalyst/dl/callbacks/checkpoint.py [hunk 2]
```diff
@@ -128,15 +128,20 @@ def load_checkpoint(*, filename, state: RunnerState):
...
 
     def get_metric(self, last_valid_metrics) -> Dict:
-        checkpoints = [
+        top_best_checkpoints = [
             (Path(filepath).stem, valid_metric)
             for (filepath, _, valid_metric) in self.top_best_metrics
         ]
-        best_valid_metrics = checkpoints[0][1]
+        all_checkpoint_metrics = [
+            (f"chekpoint_{order_index}", valid_metric)
```
Comment (Scitator) [inline catalyst/dl/callbacks/checkpoint.py]: maybe, it's better to name `epoch_{i}` metric?

File: catalyst/dl/callbacks/checkpoint.py [hunk 3]
```diff
@@ -128,15 +128,20 @@ def load_checkpoint(*, filename, state: RunnerState):
...
+            for (order_index, valid_metric) in enumerate(self.all_metrics)
+        ]
+        best_valid_metrics = top_best_checkpoints[0][1]
         metrics = OrderedDict(
             [("best", best_valid_metrics)] +
-            checkpoints +
-            [("last", last_valid_metrics)]
+            top_best_checkpoints +
+            [("last", last_valid_metrics)] +
```
Comment (Scitator) [inline catalyst/dl/callbacks/checkpoint.py:142]: btw, don't you mind to change it to
```
[("best", best_valid_metrics)] +
[("last", last_valid_metrics)] +
top_best_checkpoints +
all_checkpoint_metrics
```
?

File: catalyst/dl/callbacks/checkpoint.py (hunk 1 above)
Comment (Arquestro) [reply catalyst/dl/callbacks/checkpoint.py]: Yeah, your naming seems better. Named it relative to "best" ones.

File: catalyst/dl/callbacks/checkpoint.py (hunk 2 above)
Comment (Arquestro) [reply catalyst/dl/callbacks/checkpoint.py]: Yes, it is better one.

File: catalyst/dl/callbacks/checkpoint.py (hunk 3 above)
Comment (Arquestro) [reply catalyst/dl/callbacks/checkpoint.py:142]: Will do it.

File: catalyst/dl/callbacks/checkpoint.py (hunk 1 above)
Comment (Arquestro) [reply catalyst/dl/callbacks/checkpoint.py]: But there is `epoch_metrics` local variable in `on_epoch_end` method. I've wanted to avoid this naming.

Comment (Arquestro) [reply catalyst/dl/callbacks/checkpoint.py]: Maybe `epochs_metrics`? To suppose that we have all epochs records on main metric tracked.

Comment (Scitator) [reply catalyst/dl/callbacks/checkpoint.py]: `epochs_metrics` looks good

**Comment by TezRomacH** [comment]
@Arquestro could you add the same logic to `IterationCheckpointCallback`?

**Comment by Arquestro** [comment]
> @Arquestro could you add the same logic to `IterationCheckpointCallback`?

No problem for me, but are you sure that's okay to duplicate such logic? I mean, that it maybe doesn't fit the base class but double the code.

**Comment by Scitator** [comment]
@Arquestro For now we can try the proposal with "double code" and then check out how we can improve it. Before that – it can be overoptimization.
Maybe, we will see the general part between them and implement good solution.

**Comment by Arquestro** [comment]
Done

**Review by TezRomacH — APPROVED** [review]

**Pull request merged**